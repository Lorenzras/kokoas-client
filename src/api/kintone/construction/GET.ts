import { APPIDS, KintoneRecord } from '../config';

export const projectFields :KeyOfProjectDetails[] = [
  '$id', '$revision', 'address1',
  'address2', 'address2', 'agents',
  'buildingType', 'constructionName',
  'constructionType', 'constructionTypeId',
  'custGroupId', 'isAgentConfirmed',
  'isChkAddressKari', 'postal',
];

export const projectProspectFields: KeyOfProjectDetails[] = [
  'rank', '$id', 'schedContractDate',
  'memo', 'planApplicationDate', 'schedContractPrice',
  'estatePurchaseDate',
];

export const getConstDetails = async (recordId: string) => {
  return KintoneRecord.getRecord({
    app: APPIDS.constructionDetails,
    id: recordId,
  });
};

/**
 * Will reorganize code to put specialized api in
 * the directory of the caller.
 * @deprecated
 */
export const getFlatConstDetails = async (recordId: string) => {
  const kintoneRecord = await getConstDetails(recordId);

  return Object.entries(kintoneRecord.record).reduce((acc, [key, val]) => {

    return { ...acc, [key]: val.value };

  }, {});

};

export const getConstRecord = async (id: string) => {
  return KintoneRecord.getRecord({
    app: APPIDS.constructionDetails,
    id,
  }).then(resp => resp.record as unknown as ProjectDetails.SavedData);
};

export const getConstRecordByIds = async (ids: string[]) => {
  return KintoneRecord.getRecords({
    app: APPIDS.constructionDetails,
    query: ids
      .map(id => `${'$id' as keyof ProjectDetails.SavedData} = "${id}"`)
      .join(' or '),
  });
};

/**
 * Get projects by partial search string.
 *
 * @param search Search string
 * @returns
 */

export const searchProjects = async <
  T extends KeyOfProjectDetails,
>(search: string) => {

  return KintoneRecord.getRecords({
    app: APPIDS.constructionDetails,
    query: `${'constructionName' as T} like "${search}"`,
    fields: ['constructionName', '$id' ] as T[],
    totalCount: true,
  })
    .then(r => r.records as unknown as ProjectDetails.SavedData[]);
};