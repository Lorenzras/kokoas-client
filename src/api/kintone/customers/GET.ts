import { APP_ID } from './config';
import { KintoneRecord } from './../config';

export const getAllCustomers = () => {
  return KintoneRecord.getAllRecords({ app: APP_ID });
};

export const getCustomersByIds = (ids: string[]) => {
  const query = ids.map((id) => `$id = "${id}" `).join(' or ') + ' order by index asc';
  
  return KintoneRecord.getRecords({
    app: APP_ID,
    query,
  });
};

export const getCustomerById = (id: string) =>  KintoneRecord
  .getRecord({
    app: APP_ID,
    id,
  });


export const searchCustomers = (searchStr: string) => {
  console.log(`${'fullName' as keyof CustomerTypes.SavedData} like ${searchStr}`);
  return KintoneRecord.getRecords({
    app: APP_ID,
    query: `${'fullName'} like "${searchStr}"`,
  });
};