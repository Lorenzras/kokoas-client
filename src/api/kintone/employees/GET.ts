import { APP_ID } from './config';
import { APPIDS, KintoneRecord } from './../config';
import { resolveAffiliations, resolveRoles } from './helpers';


type KeyOfEmployee = keyof EmployeeTypes.SavedData;

export type EmployeeType = 'yumeAG' | 'cocoAG' | 'cocoConst' | 'sutekura';
export type EmpAffiliations = 'ここすも' | 'すてくら' | 'ゆめてつ';
export type EmpRoles = '店長' | '主任' | '営業' | '工務';

export const AGLabels : Record<EmployeeType, string> = {
  cocoAG : '営業担当者',
  yumeAG : 'ゆめてつAG',
  cocoConst : '工事担当者',
  sutekura: 'すてくら',
};


export interface GetEmployeesParams {
  type : EmployeeType | EmployeeType[],
  isActiveOnly?: boolean,
  storeId ?: string,
  isStoreIdRequired ?: boolean,
  territory?: '西' | '東' | null
}

export const getEmployees  = async (params ?: GetRecordParams) => {
  const queryArray = ['状態 in ("有効")', '役職 in ("店長", "主任", "営業", "工務")'];
  if (params?.query) queryArray.push(params.query);
  return KintoneRecord.getAllRecords({
    app: APP_ID,
    fields: ['$id', 'mainStoreId', 'affiliation', 'affiliateStores', '文字列＿氏名'] as KeyOfEmployee[],
    condition: queryArray.join(' and '),
  });
};


export const getCocoConst = async () => {

  return KintoneRecord.getAllRecords({
    app: APPIDS.employees,
    fields: ['$id', 'mainStore', '文字列＿氏名'] as KeyOfEmployee[],
    condition:[
      `${'状態' as KeyOfEmployee} in ("有効")`,
      `${'affiliation' as KeyOfEmployee} in ("ここすも")`,
      `${'役職' as KeyOfEmployee} in ("工務")`,
    ].join(' and '),
  }) as unknown as Promise<EmployeeTypes.SavedData[]>;
};

export const getSpecifiedEmployees = async (params: GetEmployeesParams) => {
  const {
    type = 'yumeAG',
    isActiveOnly = true,
    isStoreIdRequired = false,
    storeId,
    territory,
  } = params;

  if (isStoreIdRequired && !storeId) return [];

  const affiliations = resolveAffiliations(type)
    .map(item => `"${item}"`).join(',');

  const roles = resolveRoles(type)
    .map(item => `"${item}"`).join(',');

  return KintoneRecord.getAllRecords({
    app: APPIDS.employees,
    fields: ['$id', 'mainStore', '文字列＿氏名', 'affiliation', 'territory'] as KeyOfEmployee[],
    condition: [
      ...(isActiveOnly ? [`${'状態' as KeyOfEmployee} in ("有効")`] : []),

      ...(storeId ? [`(${'mainStoreId' as KeyOfEmployee} = "${storeId}" or ${'storeId' as KeyOfEmployee} in ("${storeId}"))`] : []),

      ...(territory ? [`${'territory' as KeyOfEmployee} = "${territory}"`] : []),

      `${'affiliation' as KeyOfEmployee} in (${affiliations})`,

      `${'役職' as KeyOfEmployee} in (${roles})`,

    ].join(' and '),
  });
};