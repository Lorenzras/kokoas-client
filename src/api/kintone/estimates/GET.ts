import { APPIDS, KintoneRecord } from '../config';


export const fetchEstimatesByProjId =  async (projId: string) => {
  if (!projId) throw new Error('Invalid project id.');

  const result = await KintoneRecord.getRecords({
    app: APPIDS.projectEstimate,
    query: `${((k: KeyOfProjEstimates)=>k)('projId')} = "${projId}"`,
  });

  return result.records as unknown as Estimates.main.SavedData[];
};

export const fetchEstimatesById = async (projEstimateId: string) => {
  const result = await KintoneRecord.getRecord({
    app: APPIDS.projectEstimate,
    id: projEstimateId,
  });

  return result.record as unknown as Estimates.main.SavedData ;
};