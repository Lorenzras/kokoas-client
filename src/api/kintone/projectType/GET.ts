import { APPIDS, KintoneRecord } from '../config';

/**
 * 工事種別名で、工事種別レコードを取得する
 *
 * @param projType 工事種別
 * @returns 工事種別のレコード
 */
export const getProjTypeByLabel = async (projType: string) => {
  const compareKeys: Array<keyof ConstructionTypes.SavedData> = ['label', 'projectName'];

  try {
    const { records } = await KintoneRecord.getRecords({
      app: APPIDS.constructionType,
      query: compareKeys
        .map(key => `${key} = "${projType}"`)
        .join(' or '),

    });

    if (records.length === 0) throw new Error('Not found.');
    if (records.length > 1) throw new Error(`${records.length} duplicates Found.`);

    return records[0] as unknown as ConstructionTypes.SavedData;

  } catch (err: any) {
    throw new Error(`Failed to get projectType. ${err.message}`);
  }
};