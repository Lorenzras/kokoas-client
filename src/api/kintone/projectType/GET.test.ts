import { getProjTypeByLabel } from './GET';

describe('Get wrapper for projectTypes', ()=>{
  it('Should get projectType by name', async ()=>{
    const result = await getProjTypeByLabel('蓄電池');

    expect(result).toMatchSnapshot();
  });
});