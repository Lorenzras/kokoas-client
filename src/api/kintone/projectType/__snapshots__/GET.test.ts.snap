// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Get wrapper for projectTypes Should get projectType by name 1`] = `
Object {
  "$id": Object {
    "type": "__ID__",
    "value": "5",
  },
  "$revision": Object {
    "type": "__REVISION__",
    "value": "3",
  },
  "description": Object {
    "type": "MULTI_LINE_TEXT",
    "value": "",
  },
  "label": Object {
    "type": "SINGLE_LINE_TEXT",
    "value": "蓄電池",
  },
  "profitRate": Object {
    "type": "NUMBER",
    "value": "",
  },
  "projectName": Object {
    "type": "SINGLE_LINE_TEXT",
    "value": "蓄電池設置工事",
  },
  "レコード番号": Object {
    "type": "RECORD_NUMBER",
    "value": "5",
  },
  "作成日時": Object {
    "type": "CREATED_TIME",
    "value": "2022-03-26T05:53:00Z",
  },
  "作成者": Object {
    "type": "CREATOR",
    "value": Object {
      "code": "RPA03",
      "name": "ラス ロレンズ",
    },
  },
  "更新日時": Object {
    "type": "UPDATED_TIME",
    "value": "2022-05-09T01:36:00Z",
  },
  "更新者": Object {
    "type": "MODIFIER",
    "value": Object {
      "code": "RPA03",
      "name": "ラス ロレンズ",
    },
  },
}
`;
